File: sections/99_master_frontier_dp_mixed/problems/b_serial_filter/solution.py
import sys
from functools import lru_cache

def count(n, pattern, target):
    digits = list(map(int, str(n)))
    p = list(map(int, pattern))
    fail = [0] * len(p)
    for i in range(1, len(p)):
        j = fail[i - 1]
        while j and p[i] != p[j]:
            j = fail[j - 1]
        if p[i] == p[j]:
            j += 1
        fail[i] = j

    def step(j, d):
        while j and p[j] != d:
            j = fail[j - 1]
        if p[j] == d:
            j += 1
        return j

    @lru_cache(None)
    def dp(pos, j, total, tight, started):
        if total > target:
            return 0
        if pos == len(digits):
            return int(total == target and (started or pattern != '0'))
        limit = digits[pos] if tight else 9
        answer = 0
        for d in range(limit + 1):
            nt = tight and d == limit
            if not started and d == 0:
                answer += dp(pos + 1, 0, total, nt, False)
            else:
                nj = step(j, d)
                if nj < len(p):
                    answer += dp(pos + 1, nj, total + d, nt, True)
        return answer
    return dp(0, 0, 0, True, False)

def main():
    data = sys.stdin.buffer.read().split()
    out = []
    for i in range(1, len(data), 3):
        out.append(str(count(int(data[i]), data[i + 1].decode(), int(data[i + 2]))))
    print('\n'.join(out))
if __name__ == '__main__':
    main()

File: sections/99_master_frontier_dp_mixed/problems/b_serial_filter/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll=long long;
string digits,pattern;
vector<int>failure;
int target;
map<tuple<int,int,int,bool,bool>,ll>memo;
int step(int j,int d) {
    char c='0'+d;
    while(j&&pattern[j]!=c)j=failure[j-1];
    if(pattern[j]==c)j++;
    return j;
}
ll dfs(int pos,int j,int sum,bool tight,bool started) {
    if(sum>target)return 0;
    if(pos==(int)digits.size())return sum==target&&(started||pattern!="0");
    auto key=tuple {
        pos,j,sum,tight,started
    }
    ;
    if(!tight&&memo.count(key))return memo[key];
    int limit=tight?digits[pos]-'0':9;
    ll answer=0;
    for(int d=0;d<=limit;d++) {
        bool next_tight=tight&&d==limit;
        if(!started&&d==0)answer+=dfs(pos+1,0,sum,next_tight,false);
        else {
            int next=step(j,d);
            if(next<(int)pattern.size())answer+=dfs(pos+1,next,sum+d,next_tight,true);
        }
    }
    if(!tight)memo[key]=answer;
    return answer;
}
ll solve(ll n,string p,int s) {
    digits=to_string(n);
    pattern=p;
    target=s;
    failure.assign(p.size(),0);
    for(int i=1;i<(int)p.size();i++) {
        int j=failure[i-1];
        while(j&&p[i]!=p[j])j=failure[j-1];
        if(p[i]==p[j])j++;
        failure[i]=j;
    }
    memo.clear();
    return dfs(0,0,0,true,false);
}
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int q;
    cin>>q;
    while(q--) {
        ll n;
        string p;
        int s;
        cin>>n>>p>>s;
        cout<<solve(n,p,s)<<'\n';
    }
}
